host-controller/src/main/java/org/jboss/as/host/controller/mgmt/HostControllerRegistrationHandler.java

```diff
@@ -332,9 +354,15 @@ public void execute(final OperationContext context, final ModelNode operation) t
             }
             // Remotely resolve the subsystem versions and create the transformation
             registrationContext.processSubsystems(transformers, extensions);
-            // Now run the read-domain model operation
-            final ReadMasterDomainModelHandler handler = new ReadMasterDomainModelHandler(hostInfo, transformers, domainController.getExtensionRegistry());
-            context.addStep(READ_DOMAIN_MODEL.getOperation(), handler, OperationContext.Stage.MODEL);
+
+            final Transformers.ResourceIgnoredTransformationRegistry ignoredTransformationRegistry;
+            final Resource resource = context.readResourceFromRoot(PathAddress.EMPTY_ADDRESS);
+            final ReadMasterDomainModelUtil.RequiredConfigurationHolder rc = hostInfo.populateRequiredConfigurationHolder(resource, domainController.getExtensionRegistry());
+            ignoredTransformationRegistry = ReadMasterDomainModelUtil.createHostIgnoredRegistry(hostInfo, rc);
+
+            final ReadMasterDomainModelUtil readUtil = ReadMasterDomainModelUtil.readMasterDomainResourcesForInitialConnect(transformers,
+                    transformationInputs, ignoredTransformationRegistry, transformationInputs.getRootResource());
+            context.getResult().set(readUtil.getDescribedResources());
```

Comment:
My guess is that you pulled this logic in here because ReadMasterDomainModelHandler calls context.acquireControllerLock(). As does ReadDomainModelHandler.

We'll need to factor this somehow so this logic only exists in one place, as it's really obscure and we'll never remember to fix any bugs in > 1 spot. Perhaps just a boolean flag to ReadMasterDomainModelHandler and ReadDomainModelHandler to turn off the context.acquireControllerLock()?